### PR TITLE
Berthing, Stage 1

### PR DESCRIPTION
**Pull request by abackwood**
Based on suggestion in #1063 

For now, a constructed berth will summon a ship automatically. The ship moves to the berth, unwrap for a moment, wraps up again and goes away. The berth then summons the next ship. This is obviously a temporary test behaviour.

Once there are no issues with this, it could maybe be integrated with trading so the ships hang around for longer and there's actually a chance for characters to interact with it.

Please test to see if any bugs arise from building and removing tiles like this.

### CODE REVIEW

File: Assets/StreamingAssets/Data/Furniture.xml [hunk 1]
```diff
@@ -453,4 +453,24 @@
...
+    <MovementCost>1</MovementCost>
+    <PathfindingModifier>1</PathfindingModifier>
+    <PathfindingWeight>0</PathfindingWeight>
+    <Width>3</Width>
+    <Height>1</Height>
+
+    <BuildingJob jobTime="1" />
+
+    <Params>
+      <Param name="occupied" value="0" />
```
Comment (zwrawr) [inline Assets/StreamingAssets/Data/Furniture.xml]: In your .xml files changes , some use 2 spaces indent some are 4 spaces and some are tabs. You should probably pick one

**Comment by Grenkin1988** [comment]
I've placed this like this.
image

Ship came and took my floor away. I need my floor :-(

**Comment by abackwood** [comment]
@Grenkin1988 Yeah my code doesn't check for that yet. You may have noticed a bunch of errors popping up telling you the floor was actually occupied and that's bad?

@zwrawr I'm not sure what you mean. Those lines are all indented by 2 spaces, just like the rest of the file? Nested tags are obviously indented further.

**Comment by mikejbrown** [comment]
Why is this touching 86 files??!

**Comment by mikejbrown** [comment]
Travis errors:
```
Build Failed! 

The compiler generated the following messages:

-----CompilerOutput:-stdout--exitcode: 1--compilationhadfailure: True--outfile: Temp/Assembly-CSharp.dll

Compilation failed: 1 error(s), 0 warnings

-----CompilerOutput:-stderr----------

Assets/Scripts/Models/Prototypes/ShipPrototypes.cs(32,9): error CS0103: The name `SetPrototype' does not exist in the current context

-----EndCompilerOutput---------------
```
This was probably broken by #1146.

File: Assets/Scripts/Models/Prototypes/ShipPrototypes.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        try
+        {
+            ship.ReadXmlPrototype(reader);
+        }
+        catch (Exception e)
+        {
+            LogPrototypeError(e, ship.ShipType);
+        }
+
+        SetPrototype(ship.ShipType, ship);
```
Comment (mikejbrown) [inline Assets/Scripts/Models/Prototypes/ShipPrototypes.cs]: Yes, it was broken by #1146. You need `Set` here.

**Comment by koosemose** [comment]
Conflicted.

You may also want to look into making sure it doesn't take tiles around the berth (possibly just prevent landing at a berth with tiles or furniture occupying what will be ship's space. Perhaps also check if it's in the outside room to function.

**Comment by GamerInspired** [comment]
I think it shouldn't allow ships to come if there is an object in the way; so if there is furniture in front of it but maybe still allow them arrive if there is a floor; maybe it landed on the floor?

**Comment by abackwood** [comment]
The berth being outside isn't technically a requirement. If the ship can pathfind to it and there is enough space I don't see why we'd disallow it. Note: the ships dont pathfind yet.

@GamerInspired: I'm going to make it check for enough space but it wouldnt be landing on the floor. The kind of ship that berths are large container ships and passenger liners, the way i picture it.
Interesting thought for later though: the existing landing pad could be a place for smaller craft to sit. Drones and fighters and the like. Those don't need a full berth.

**Comment by koosemose** [comment]
Once they pathfind, and once pathfinding properly handles Z-levels it shouldn't be able to pathfind to somewhere that's not outside (assuming pathfinding can be told not to open doors), so it seems like it would be useful to not call a ship or at least skip pathfinding if there is an easy check for a situation in which it wouldn't be able to pathfind to.

**Comment by GamerInspired** [comment]
Just thought for the far future that bay doors would be awesome; to let small to medium ships in and out while also protecting them.

**Comment by abackwood** [comment]
Resolved conflicts. WIP until I can verify that everything still works. Of ourse others can feel free to do the same if you get there before I do. Could be most of tomorrow before I get to it :)

File: Assets/Scripts/Controllers/ModsManager.cs [hunk 3]
```diff
@@ -37,8 +37,11 @@ public void LoadPrototypes()
         PrototypeManager.Need.LoadPrototypes(mods);
         PrototypeManager.Trader.LoadPrototypes(mods);
         PrototypeManager.SchedulerEvent.LoadPrototypes(mods);
+        PrototypeManager.Ship.LoadPrototypes(mods);
+
         PrototypeManager.Stat.LoadPrototypes(mods);
         PrototypeManager.Quest.LoadPrototypes(mods);
+        PrototypeManager.Ship.LoadPrototypes(mods);
```
Comment (sboigelot) [inline Assets/Scripts/Controllers/ModsManager.cs]: PrototypeManager.Ship.LoadPrototypes(mods);
PrototypeManager.Ship.LoadPrototypes(mods);

Duplicate?

**Comment by mikejbrown** [comment]
The test behavior seems to work as described but it spams errors (either `Tile 44,54 is not empty. Replacing anyway.` etc. or `Tried to delete the outside room.` depending on whether the berth is built indoors or out).
Also this pull request has Stylecop violations. Please fix these and remove [Stylecop violations] from the title.
```
C:\Users\Michael\Documents\Github\ProjectPorcupine\Assets\Scripts\Models\Character.cs(1,1): Warning: SA1202 : CSharp.Ordering : All private methods must be placed after all public methods. (Assembly-CSharp)
C:\Users\Michael\Documents\Github\ProjectPorcupine\Assets\Scripts\Models\World.cs(1,1): Warning: SA1513 : CSharp.Layout : Statements or elements wrapped in curly brackets must be followed by a blank line. (Assembly-CSharp)
```

**Comment by abackwood** [comment]
Should I remove the "tile occupied" error? It signifies exactly what it sounds like: that the ship is trying to berth over some existing stuff which is bad. I think that's worth pointing out.

The "deleting outside room" error is something that I don't know what causes it. Sounds like something to do with rooms and floodfill. Changing that here might change a bunch more files that shouldnt even have anything to do with ships. Do you really want me to shoehorn that in here?

I didnt change Character.cs so that wasnt caused by me. I'll take a look at the World.cs one.

**Comment by vogonistic** [comment]
I don't think the ship should arrive if there isn't enough space for it. That would be an easy check before. Maybe for the first pass off this functionality, it would only be allowed in open space with no floors either.

**Comment by mikejbrown** [comment]
This pull request has merge conflicts. Please resolve the conflicts and remove [Conflicts] from the title.

**Comment by mikejbrown** [comment]
Is there any movement on this? If the only problem is merge conflicts and you don't have the time to fix it you can probably [enable maintainers to make changes](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) to this and one of the GKs will (eventually) do it for you.

**Comment by abackwood** [comment]
Sorry ive been leaving this. Yeah it really is just a relatively minor change and solving merge conflicts.

I'll try and find time today, otherwise I'll enable maintainers so soneone else can fix it to not broken and merge it.

**Comment by abackwood** [comment]
Resolved and added above-mentioned functionality of refusing to berth where there is no space.

* I don't think the refusal to berth works if you place walls/floors in its way while it's already coming in. I figured for now this is fair enough and we'll make ships actually 'intelligent' later.
* Furniture on the ship is deconstructed when it wraps ("deberths"). This now leaves raw materials. Obviously this needs to be fixed but with this PR already so old I hope this will be admissible for now. I don't know anything about how dropping materials on deconstruction is done so it would take a while for me to five into this. It doesn't break anything regardless.

**Pull request merged**